src/Compilers/CSharp/Test/Semantic/Semantics/RecordTests.cs

```diff
@@ -426,5 +430,444 @@ public void EmptyRecord()
+        public void WithExpr8()
+        {
+            var src = @"
+class B
+{
+    public int X { get; }
+    public virtual B With(int X) => null;
+}
+class C : B
+{
+    public string Y { get; }
+    public static void Main()
+    {
+        var c = new C();
+        B b = c;
+        b = b with { };
+        c = c with { };
+    }
+    public override B With(int X) => null;
+    public virtual C With(string Y) => null;
+}";
+            var comp = CreateCompilation(src);
+            comp.VerifyDiagnostics();
+        }
+
+        [Fact]
+        public void WithExpr9()
+        {
+            var src = @"
+class C
+{
+    public int X { get; }
+    public string With(int X) => null;
+    public static void Main()
+    {
+        var c = new C();
+        c = c with { };
+    }
+}";
+            var comp = CreateCompilation(src);
+            comp.VerifyDiagnostics(
+                // (9,13): error CS8804: The type of the 'with' expression receiver, 'C', does not derive from the return type of the 'With' method, 'string'.
+                //         c = c with { };
+                Diagnostic(ErrorCode.ERR_ContainingTypeMustDeriveFromWithReturnType, "c").WithArguments("C", "string").WithLocation(9, 13),
+                // (9,13): error CS0029: Cannot implicitly convert type 'string' to 'C'
+                //         c = c with { };
+                Diagnostic(ErrorCode.ERR_NoImplicitConv, "c with { }").WithArguments("string", "C").WithLocation(9, 13)
+            );
+        }
+ 
+        [Fact]
+        public void WithExpr10()
+        {
+            var src = @"
+class C
+{
+    public int X { get; }
+    public string With(int X) => null;
+    public static void Main()
+    {
+        var c = new C();
+        c = c with { };
+    }
+}";
+            var comp = CreateCompilation(src);
+            comp.VerifyDiagnostics(
+                // (9,13): error CS8804: The type of the 'with' expression receiver, 'C', does not derive from the return type of the 'With' method, 'string'.
+                //         c = c with { };
+                Diagnostic(ErrorCode.ERR_ContainingTypeMustDeriveFromWithReturnType, "c").WithArguments("C", "string").WithLocation(9, 13),
+                // (9,13): error CS0029: Cannot implicitly convert type 'string' to 'C'
+                //         c = c with { };
+                Diagnostic(ErrorCode.ERR_NoImplicitConv, "c with { }").WithArguments("string", "C").WithLocation(9, 13)
+            );
+        }   
+ 
+        [Fact]
+        public void WithExpr11()
+        {
+            var src = @"
+class C
+{
+    public int X { get; }
+    public C With(string X) => null;
+    public static void Main()
+    {
+        var c = new C();
+        c = c with { X = """"};
+    }
+}";
+            var comp = CreateCompilation(src);
+            comp.VerifyDiagnostics(
+                // (9,13): error CS8806: The 'With' method parameter named 'X' has type 'string' which doesn't match member type 'int'
```

Comment:
This is unfortunate error message placement: the error is on the with method, not the usage. Could we consider flagging both the usage location and the definition?